### ISSUE TITLE
Consider shas of archive contents as opposed to archive itself for downloads

### ISSUE THREAD

**Issue by keith** [bazelbuild/bazel#19033]
### Description of the feature request:

One still mostly unsolved problem in the community is that github tars are mostly not guaranteed to be stable (even though they change very rarely). bzlmod is trying to enforce this but not all repos create a stable archive for their releases today, so I don't see this problem going away. One thing I noticed testing out nix is that they use the combined sha of the contents of the archive instead of the sha of the archive itself, which I imagine would be stable against this issue.

Related slack discussion: https://bazelbuild.slack.com/archives/CA31HN1T3/p1690214404391649

**Comment by meisterT** [comment]
related: https://blog.bazel.build/2023/02/15/github-archive-checksum.html

**Comment by Wyverald** [comment]
See also various public meltdowns regarding this very topic, last time this happened:
* https://github.com/orgs/community/discussions/45830#discussioncomment-4824814
* https://github.com/orgs/community/discussions/45830#discussioncomment-4824881
* https://github.com/orgs/community/discussions/46034#discussioncomment-4844276

**Comment by fmeum** [comment]
I went ahead and made some minor tweaks to Go's implementation of content hashing for ZIP files to produce a standalone tool that unzips ZIP files while checking their content hash: https://github.com/fmeum/content_hash_unzip

The tool places a number of restrictions on the contents: symlinks are not allowed, all files are implicitly executable, no file metadata is preserved, ...

Please let me know if you find this to break on your favourite ruleset.

**Comment by meteorcloudy** [comment]
Thanks for doing the experiment!

> The tool places a number of restrictions on the contents

Those restrictions sound pretty intrusive, but I guess they can be solved?

Besides, I can see a few additional complexities:

- We'll need to embed binaries of this tool for different platforms, which increases the Bazel binary size.
- We'll actually need to fetch the go toolchains for building Bazel.
- A more general problem, Will content-based hash work with the current "repository cache" (download artifact cache), can it be safely used as the key to fetch the cache? If not, does that mean we still need the sha hash to make the caching work?

Lastly, I think I forgot the details about the security implication of this. Why is the go implementation safer than a potential native [java implementation](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java;l=33?q=compress&sq=&ss=bazel)?

**Comment by fmeum** [comment]
> Those restrictions sound pretty intrusive, but I guess they can be solved?

They would be very intrusive for arbitrary archives, but I expect them to be less problematic for the use case of archives generated automatically from Git repositories. We can of course extend this to something like nix's NAR format, which has fewer restrictions but also, as far as I know, no existing cross-platform implementation that's already proven to work well in a security-sensitive context.

> * We'll need to embed binaries of this tool for different platforms, which increases the Bazel binary size.

Wouldn't we only have to ship the version build for the same platform as Bazel? This would only ever run during the execution of a repo rule. I just checked that the binary compresses down to less than 800 KiB (1.7 MiB uncompressed). This could likely be optimized further.

> * A more general problem, Will content-based hash work with the current "repository cache" (download artifact cache), can it be safely used as the key to fetch the cache? If not, does that mean we still need the sha hash to make the caching work?

This effort only really makes sense if we end up trusting it to an extent where we allow the content-based hash to be used as a key in the "downloaded artifact" aka remote cache. This support wouldn't be automatic, we would need to add it.

> Lastly, I think I forgot the details about the security implication of this. Why is the go implementation safer than a potential native [java implementation](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java;l=33?q=compress&sq=&ss=bazel)?

It's not necessarily safer, but it's more obviously safe: This tool is essentially just a copy of the `go mod` code with very minor modifications that probably wouldn't be too difficult to review. By basing this on an existing, maintained and likely security-reviewed implementation, we would reduce the cost of maintaining this feature.

We could of course go ahead and implement a Java port of this, but then we would be on our own in terms of ensuring that it actually fulfills its intended security guarantees. This would ultimately come down to the question of who would be willing to maintain it long-term.

**Comment by chandlerc** [comment]
Just wanted to surface that we've started to see intermittent failures that look very much like another potential change to the sha256.

For example this run: https://github.com/carbon-language/carbon-lang/actions/runs/22159595633/job/64081143764?pr=6518

When we download https://github.com/llvm/llvm-project/archive/21a1e6e6a70d70635c68b9aaa54f816ae36a6416.tar.gz we are seeing a sha256 of (base64) ouWG2d4Cob9P6+dxCLZnV4RldSOqZp7mtwd6vd+ZK3Q= locally but cmdJ+wsi2Nuox86fldQjf5zlc7+aEQGdrM98Kx061j0= on an action runner.

Any progress on moving to a content-hashing approach?

We're worried that there may be another change being piloted, or rolled out with little warning and causing similar broad breakage as before.

**Comment by meteorcloudy** [comment]
We are not planning this feature right now, there are still many unknown.

Are you still seeing this mismatch in the latest build? My feeling this isn't a change from github side, otherwise we should have seen major disruption across different ecosystems instead of just one URL.

**Comment by chandlerc** [comment]
> Are you still seeing this mismatch in the latest build? My feeling this isn't a change from github side, otherwise we should have seen major disruption across different ecosystems instead of just one URL.

We keep seeing periodic failures:
https://github.com/carbon-language/carbon-lang/actions/runs/22785819172/job/66102267016

It really looks like a canary or 1% test deployment which causes a rotation... I am suspecting that we're just the first to notice, and the major disruption is coming. =]

We're currently moving off `http_archive` and to `git_override` as this keeps breaking us. But my understanding is that is fundamentally less efficient for Bazel, and I'm worried about the wider ecosystem having to do that which is why I wanted to bump this up.

**Comment by meteorcloudy** [comment]
> It really looks like a canary or 1% test deployment which causes a rotation... 

Yeah, that would be concerning. But hopefully github keeps their promise to give a [six months’ notice](https://github.blog/open-source/git/update-on-the-future-stability-of-source-code-archives-and-hashes/#future-stability-of-archives-and-hashes).

The best way to workaround is probably to use a mirror. We already have this for BCR modules (`--module_mirrors=https://bcr.cloudflaremirrors.com`), but you'll still need it for your custom http_archive.

The content hash solution should probably be the long term solution, but it'll take some time to design, implement and migrate.
